L1Trigger/L1TGlobal/src/GlobalBoard.cc

```diff
@@ -1180,23 +1182,45 @@ const std::vector<double> l1t::GlobalBoard::semirandomNumber(const edm::Event& i
   std::srand(std::rand() + iEvent.id().event());
   // very large (semi)random number
   double const semirandom = std::rand();
-  for (auto& ps : out) {
-    // if the ps is smaller than 1 (e.g. ps=0, ps=1), it is not changed
-    // else, replace ps with a semirandom integer in the [1,ps] range
-    if (ps > 1) {
-      auto nps = semirandom - floor(semirandom / ps) * ps;
-      // if nps=0 or a wrong value (<0,>ps) use PS value (standard method)
-      if (nps > 0 and nps <= ps)
-        ps = nps;
-      else {
-        if (nps != 0)  // complain only if nps <0 or nps >PS
-          edm::LogWarning("L1TGlobal::semirandomNumber")
-              << "\n The inital prescale counter obtained by L1TGlobal::semirandomNumber is wrong."
-              << "\n This is probably do to the floating-point precision. Using the PS value."
-              << "\n semirandom = " << semirandom << "\n PS = " << ps << "\n nps = " << nps
-              << " <-- it should be in the range [0 , " << ps << "]" << std::endl;
-      }
+
+  for (size_t iAlgo = 0; iAlgo < prescaleFactorsAlgoTrig.size(); iAlgo++) {
+    int ps = std::lround(prescaleFactorsAlgoTrig[iAlgo] * std::pow(10, m_precision));
+    int nps = semirandom - floor(semirandom / ps) * ps;
+
+    // if nps=ps or a wrong value (<0,>ps) use zero
+    if (nps >= 0 and nps < ps)
+      out.push_back(PrescaleCounter(prescaleFactorsAlgoTrig[iAlgo], m_precision, nps));
```

Comment:
I'm not sure I understand this. Assume `prescaleFactorsAlgoTrig[iAlgo]` is `1.35`; if `semirandom` is `50`, then `nps` will be `50`. What happens then, what's the effective prescale being used? Instead (trying to guess the intent of this function), I would have expected it to return a prescale with a random value `X` with `1.00 < X <= 1.35`.

Reply:
The way the fractional prescales are implemented in firmware, the prescale is multiplied by 10^precision (= `prescale_count` in the `PrescaleCounter` struct) and then the counter (`trigger_counter` in the `PrescaleCounter` struct) is incremented by 10^precision (= `single_step` in the `PrescaleCounter` struct) until we are > prescale_count. The semirandomNumber function is supposed to set the `trigger_counter` to a random number, so that it doesn't always start from 0. So we want this function to construct `PrescaleCounter` objects with `trigger_counter` initially between 0 and prescale_count - 1.